### PR TITLE
Generate a .ycm_extra_conf.py file for use with YouCompleteMe

### PR DESCRIPTION
**Pull request by dalg24**
Configure with `-D DataTransferKit_ENABLE_YouCompleteMe=ON` to generate the .ycm_extra_conf.py file in the root directory.

### CODE REVIEW

File: cmake/.ycm_extra_conf.py.in [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    # ycm_extra_conf IF YOU'RE NOT 100% SURE YOU NEED IT.
+    try:
+      final_flags.remove( '-stdlib=libc++' )
+    except ValueError:
+      pass
+  else:
+    relative_to = DirectoryOfThisScript()
+    final_flags = MakeRelativePathsInFlagsAbsolute( flags, relative_to )
+
+  return { 'flags': final_flags }
```
Comment (aprokop) [inline cmake/.ycm_extra_conf.py.in]: This line does not work for me. We may have different versions. My default one (from ycmd) instead has
```
  return {
    'flags': final_flags,
    'do_cache': True
  }
```
If I don't have those line, I get an error: ```KeyError: u'do_cache```

Comment (dalg24) [inline cmake/.ycm_extra_conf.py.in]: Mmm.  You might need to update it.  Migrating to Vundle might solve the problem.

**Comment by dalg24** [comment]
@Rombur would you check if that works for you?

**Comment by Rombur** [comment]
Doesn't work for me.

**Comment by Rombur** [comment]
OK silly mistake. It works

**Comment by aprokop** [comment]
OK. It works for me for regular completions. But ```:YouCompleteMe GoTo``` still does not seem to work. It may be just the case that you need extra stuff for it.

**Comment by dalg24** [comment]
@aprokop @Rombur I added a `.vimrc` file in the Docker image, let me know what you would like to add to it.

File: cmake/.ycm_extra_conf.py.in [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+import ycm_core
+
+# These are the compilation flags that will be used in case there's no
+# compilation database set (by default, one is not set).
+# CHANGE THIS LIST OF FLAGS. YES, THIS IS THE DROID YOU HAVE BEEN LOOKING FOR.
+flags = [
+'-Wall',
+'-Wextra',
+'-Werror',
+'-Wc++98-compat',
```
Comment (Rombur) [inline cmake/.ycm_extra_conf.py.in]: Get ride of this. Kokkos requires C++11.

**Comment by aprokop** [comment]
First, I'd recommend installing
```
Plugin 'embear/vim-localvimrc
```
and renaming `.vimrc` to `.lvimrc`. This way vim will automatically pick it up, but only for this project.
Second, here is what my `.lvimrc` looks right now for highlighting class names and macros:
```
set ts=8

syn keyword cDefine
    \ DTK_CHECK
    \ DTK_CHECK_ERROR_CODE
    \ DTK_ENSURE
    \ DTK_INSIST
    \ DTK_REMEMBER
    \ DTK_REQUIRE

" DTK
syn keyword cType
    \ BasicEntitySet
    \ BasicEntitySetIterator
    \ BasicGeometryEntity
    \ BasicGeometryEntityImpl
    \ BasicGeometryExtraData
    \ BasicGeometryLocalMap
    \ BasicGeometryManager
    \ BoxGeometry
    \ BoxGeometryImpl
    \ BuhmannBasis
    \ CenterDistributor
    \ ClientManager
    \ CloudDomain
    \ CoarseGlobalSearch
    \ CoarseLocalSearch
    \ ConsistentInterpolationOperator
    \ CylinderGeometry
    \ CylinderGeometryImpl
    \ DataTransferKitException
    \ Entity
    \ EntityId
    \ EntityCenteredField
    \ EntityCenteredShapeFunction
    \ EntityExtraData
    \ EntityImpl
    \ EntityIntegrationRule
    \ EntityIterator
    \ EntityLocalMap
    \ EntitySet
    \ EntityShapeFunction
    \ EuclideanDistance
    \ Field
    \ FineLocalSearch
    \ FunctionSpace
    \ IntrepidBasisFactory
    \ IntrepidCell
    \ IntrepidIntegrationRule
    \ IntrepidShapeFunction
    \ IntrepidSideCell
    \ L2ProjectionOperator
    \ LocalEntityPredicate
    \ LocalMLSProblem
    \ MapOperator
    \ MapOperatorFactory
    \ MovingLeastSquareReconstructionOperator
    \ NanoflannTree
    \ NewtonSolver
    \ NodeToNodeOperator
    \ NonlinearProblemTraits
    \ ParallelSearch
    \ Point
    \ PointCloudOperatorFactory
    \ PointInFaceVolumeOfInfluenceNonlinearProblem
    \ PolynomialMatrix
    \ PredicateFunction
    \ ProjectionPrimitives
    \ ProjectPointFeatureToEdgeFeatureNonlinearProblem
    \ ProjectPointToFaceNonlinearProblem
    \ RadialBasisPolicy
    \ SearchTreeFactory
    \ SplineCoefficientMatrix
    \ SplineEvaluationMatrix
    \ SplineInterpolationOperator
    \ SplineInterpolationPairing
    \ SplineProlongationOperator
    \ SupportId
    \ TpetraMap
    \ TpetraVector
    \ TpetraMultiVector
    \ UndefinedNonlinearProblemTraits
    \ UndefinedRadialBasisPolicy
    \ WendlandBasis
    \ WuBasis

" Trilinos Types
syn keyword cType
    \ GO GlobalOrdinal
    \ LO LocalOrdinal
    \ NO Node
    \ SC Scalar

" Teuchos
syn keyword cType
    \ Array
    \ ArrayRCP
    \ ArrayView
    \ ParameterList

" Tpetra
syn keyword cType
    \ CrsMatrix
    \ Distributor
    \ Import
    \ Map
    \ MultiVector
    \ Vector

" Intrepid
syn keyword cType
    \ IntrepidcellLocalMap

" STK
syn keyword cType
    \ STKMeshEntity
    \ STKMeshEntityExtraData
    \ STKMeshEntityImpl
    \ STKMeshEntityIntegrationRule
    \ STKMeshEntityIterator
    \ STKMeshEntityIteratorRange
    \ STKMeshEntityLocalMap
    \ STKMeshEntitySet
    \ STKMeshField
    \ STKMeshHelpers
    \ STKMeshManager
    \ STKMeshNodalShapeFunction
    \ STKPartNamePredicate
    \ STKPartPredicate
    \ STKPartVectorPredicate
    \ STKSelectorPredicate
```

**Comment by aprokop** [comment]
Are we ready to merge?

**Comment by dalg24** [comment]
Do you or @Rombur want to add anything to `.vimrc`?

**Comment by dalg24** [comment]
If so just push new commits onto that branch.  I will take care of the rebase.

**Comment by Rombur** [comment]
It's fine for me.

**Comment by aprokop** [comment]
Go ahead and rebase/merge.

**Comment by dalg24** [comment]
```
root@87d1dd1a276e:/# du -skh /root/.vim
428M	/root/.vim
```
ouch!

**Comment by dalg24** [comment]
Ready for final review.  Note that I had made a typo in the `docker-compose.yml` file.  I was pointing to some local image instead of the one from Docker Hub.

**Pull request merged**